### ISSUE TITLE
Non strict mode

### ISSUE THREAD

**Issue by 89z** [ohler55/ojg#76]
I am trying to parse JSON from some input like this:

~~~js
window.__additionalDataLoaded('extra',{"shortcode_media":999});
~~~

Where the output would be:

~~~json
{"shortcode_media":999}
~~~

is something like that possible? Similar to `Strict = false` with XML:

https://godocs.io/encoding/xml#Decoder.Strict

**Comment by ohler55** [comment]
XML has a lot more encoding rules so have a non-strict mode makes sense. JSON is much simpler to there is generally not a need to have a non-strict mode unless you go full on lazy like the SEN format of OjG at https://github.com/ohler55/ojg/blob/develop/sen.md. Oj for Ruby does not support that but the example you gave looks like a Hash and not a JSON string. What is the input JSON you would like to parse?

**Comment by 89z** [comment]
> What is the JSON string that you would like to parse?

I dont understand the question. The original input looks exactly like the first example, exception with more keys in the object. The ideal output would be substring thats just the JSON part, or maybe a slice of all valid JSON substrings. Or maybe even a way to unmarshal directly from a non strict input. I dont know the best approach, I am just trying to see what options are available, if any.

**Comment by ohler55** [comment]
The original looks like a functions call on some variable named `window`.  The syntax for the `{"shortcode_media":999}` is odd in that is appear to be an attempt as a map but there is not `map[string]interface{}` in front of it. Ar you saying the whole line is a string like this: ```"window.__additionalDataLoaded('extra',{\"shortcode_media\":999})"```?

You are using golang, right?

**Comment by 89z** [comment]
> 
> 
> Ar you saying the whole line is a string like this: `"window.__additionalDataLoaded('extra',{\"shortcode_media\":999})"`?

Yes.

> 
> You are using golang, right?

Yes.

**Comment by ohler55** [comment]
One thing I've run into is trying to parse mongo query output which usually has a few lines of non JSON. It seems like you might be looking for something similar. Basically ignore the leading non JSON as well as any trailing non JSON. Does that sum up what you are looking for?

**Comment by ohler55** [comment]
If you are interested in extracting JSON from a string or stream with other non JSON characters then OjG does not currently support that. It seems like a great topic for discussion though for an enhancement. Would you like to start a Discussion (GitHub discussion) and help formulate what the feature might look like?

**Comment by 89z** [comment]
> One thing I've run into is trying to parse mongo query output which usually has a few lines of non JSON. It seems like you might be looking for something similar. Basically ignore the leading non JSON as well as any trailing non JSON. Does that sum up what you are looking for?

Pretty much yes. Possibly, in some cases it could be multiple runs of JSON in a single string, like:

~~~
<GARBAGE> <JSON> <GARBAGE> <JSON> <GARBAGE>
~~~

but I would guess thats rare. More common would be what you said, just a single run of JSON:

~~~
<GARBAGE> <JSON> <GARBAGE>
~~~

> It seems like a great topic for discussion though for an enhancement. Would you like to start a Discussion (GitHub discussion) and help formulate what the feature might look like?

I have been searching for a solution to this problem for many hours and days. I have seen many RegExp approaches, but those are pretty weak if the surrounding context changes. Then, I thought about finding or implementing a JavaScript parser, so I could extract the JSON from that. However after trying several approaches, it just seems like overkill for the problem, as a JavaScript parser is several thousand lines minimum, and all the packages I found had one problem or another. So this morning I had the idea, that tackling from the JSON side might be the better way to go. I even wrote some code that might work [1], if you want to look at it.

I am happy to add to a discussion, but I just dont have the energy to start something like that. Hope you can understand.

1. https://github.com/89z/parse/blob/master/json/json.go

**Comment by ohler55** [comment]
Discussion started here: https://github.com/ohler55/ojg/discussions/78

All I ask is to participate in the discussion. I'd be glad to add the feature once we get it described to a consensus.

**Comment by ohler55** [comment]
@89z is there some reason you removed all the comments and the issue title?

**Comment by 89z** [comment]
Sorry for the trouble, a user was making threats against me, so for my own protection:

~~~
https://github.com/tidwall
~~~

**Comment by ohler55** [comment]
Wow, that's crazy. You had some good ideas that I intend to follow through on. If there is anything I can do to help let me know.

**Comment by 89z** [comment]
The user in question has several repos related to JSON (at least 8 I think). On accident, I posted the same issue to two of his repos. He did not like this, so he took it upon himself to look through my personal repos and other recent issues Ive posted in other places, even going so far as to reading entire threads involving me, but having nothing to do with him.

I dont know about you, but I did not like this behavior at all, it made me quite uncomfortable. I am free to post issues as I like, if owner like the issue, then we can work to solve it. If owner dont like the issue, then they can close it. But this was beyond that, and not appropriate behavior in my opinion. When I confronted this person about their behavior, they got defensive, and threaten to report me to moderator. I dont have malice when I am posting issues, I am just looking for solution. If repo owner dont like it, just ignore the issue or close it, simple. No need for weird behavior this person did.

In regard to the issue on this page, I am still willing to converse, if you decide to move forward with some implementation.
